### ISSUE TITLE
libsrshim posts corrupt notifications: likely mis-encoding of oldname field

### ISSUE THREAD

**Issue by petersilva** [MetPX/sarrac#36]
The HPC mirroring client (python subscribers) are getting stuck on an apparent misencoding of the oldname field.  In the sr_subscribe log:

```
2019-05-03 12:39:15,852 [INFO] file_age 22367.3 seconds for /space/hall2/sitestore/eccc/cmod/cmoi/smco501/ldpreload_test/products/products_dbase/iweb/par/prog/global/2019050300/2019050300_054_G1_west_I_EXPERIMENTAL_uv@sfc@pn@pubwarn_078.png
2019-05-03 12:39:15,854 [ERROR] sr_amqp/consume Type: <class 'UnicodeDecodeError'>, Value: 'utf-8' codec can't decode byte 0x80 in position 0: invalid start byte
2019-05-03 12:39:15,854 [ERROR] Could not consume in queue q_MSC-SCI-CMC.sr_subscribe.mirror-ss1-from-hall2-p
```

**Comment by petersilva** [comment]
String values are supposed to be utf-8 data.  When I looked in the rabbitmq management gui, I used get messages in the queue do see some samples.  Each one had the oldname: field was corrupt, reporting utf-8 decoding errors.  It looks like a deep directory was renamed, and there is some bug in encoding the oldname field.   To reproduce exactly the issue, would try:   start a shell with the shim library active, mv a deep directory, and have a subscriber consuming the posts.

**Comment by petersilva** [comment]
Could we have @msaraga try to reproduce this?

**Comment by petersilva** [comment]
@msaraga and I just looked over the code together.  The C just assumes the filenames are utf-8, does no conversion or validation.  It could be that someone has an iso8859-1 or other encoded file name, and copying those bytes results in an invalid utf-8 string.

**Comment by petersilva** [comment]
ya know,  we could do something much simpler... we just validate that it is utf8 encoding, and, ifnot we strip it until it is... we look for the two higher order bits, if there is 11 in top two bits, then we look at the next one, if it is either 11 or 10, then we continue... if the high order bits don´t follow utf-8, then we can either drop the byte, or clear the high bit... not sure which is less trouble, we can emit a warning message saying that we found a non utf-8 character and are mangling it.   We don´t have to worry about other character sets or using a real charset library, we just say your file names better be utf8, or else.

**Comment by petersilva** [comment]
so no additional dependency, just a check for make_sure_string_is_unicode, somehow...

**Comment by msaraga** [comment]
on behalf of @petersilva :

so the field that is always corrupt is oldname. One should look at how values of oldname get set.
There is a strdup() used to set it in sr_post.c, might be good to confirm the return value is non-null
before assigning it.

**Comment by petersilva** [comment]
Another possibility is that the client mentioned the presence of % character in the name is messing things up... I don´t that should cause a utf-8 decode errors though.

**Comment by petersilva** [comment]
OK, three occurrences in the past week.  This is now a severe problem.

* need a work-around that does something to prevent corruption... either truncating or the field or eliding it from the post.

ideally we find the actual cause and correct.

**Comment by msaraga** [comment]
#39

**Comment by msaraga** [comment]
while testing functionality for a feature, it was found that deleting a directory containing a file with `rm -rf dir_name` resulted in the following output:
```
aspymis@amqpbot:~/cpost_test$ sr_cpost -c cpost_test.conf -sum 0
2019-05-23 13:01:49,380 [INFO] cpost 2.19.05b1ubuntu2 config: cpost_test, pid: 17553, starting
2019-05-23 13:02:11,881 [INFO] published: 20190523130211.880591284 file:/ /home/aspymis/cpost_test/dd/s/t topic=v02.post.home.aspymis.cpost_test.dd.s sum=0,958 source=bunnymaster to_clusters=localhost from_cluster=localhost mtime=20190523130211.876 atime=20190523130211.876 mode=0664 parts=1,0,1,0,0
2019-05-23 13:02:18,494 [INFO] published: 20190523130218.494300923 file:/ /home/aspymis/cpost_test/dd/s/t topic=v02.post.home.aspymis.cpost_test.dd.s sum=R,99f97d455d5d62b24f3a942a1abc3fa8863fc0ce2037f52f09bd785b22b800d4f2e7b2b614cb600ffc2a4fe24679845b24886d69bb776fcfa46e54d188889c6f source=bunnymaster to_clusters=localhost from_cluster=localhost
2019-05-23 13:02:18,495 [ERROR] file path "/home/aspymis/cpost_test/dd/s/" not utf8 encoded, ignoring sr_post call
2019-05-23 13:02:18,496 [INFO] published: 20190523130218.495899027 file:/ /home/aspymis/cpost_test/dd/s topic=v02.post.home.aspymis.cpost_test.dd sum=R,2c1ee68372215b1ce064426b5cdbd4ef2581ace0dd3b21fa2be27f364827242e83f68b68be03f5b3e24be5d1b4315f98a0a96d19713fb3a19dc455fb6adc3431 source=bunnymaster to_clusters=localhost from_cluster=localhost
```

cpost_test.conf:
```
post_broker amqp://bunnymaster@localhost/
post_exchange xtest

path /home/aspymis/cpost_test/dd

sleep 1

post_base_url file:/

events create,delete,link,modify

accept .*
```

Command log:
```
aspymis@amqpbot:~/cpost_test/dd$ mkdir s
aspymis@amqpbot:~/cpost_test/dd$ touch s/t
aspymis@amqpbot:~/cpost_test/dd$ rm -rf s
```

**Comment by petersilva** [comment]
great! no we can reproduce it, need to figure out how to fix it.

**Comment by msaraga** [comment]
Same case as above, with the sr_cpost running in `debug` mode:

```
aspymis@amqpbot:~/cpost_test$ sr_cpost --debug -c cpost_test.conf 
2019-05-24 18:19:46,378 [DEBUG] argument after substitutions: cpost_test.conf
2019-05-24 18:19:46,380 [DEBUG] sr_config_read 1 failed to open: /home/aspymis/.config/sarra/cpost/cpost_test.conf
2019-05-24 18:19:46,381 [DEBUG] sr_config_read 2 failed to open: /home/aspymis/.config/sarra/cpost/cpost_test
2019-05-24 18:19:46,382 [DEBUG] sr_config_read trying to open: cpost_test.conf
2019-05-24 18:19:46,383 [DEBUG] argument after substitutions: amqp://bunnymaster@localhost/
2019-05-24 18:19:46,384 [DEBUG] argument after substitutions: xtest
2019-05-24 18:19:46,385 [DEBUG] argument after substitutions: /home/aspymis/cpost_test/dd
2019-05-24 18:19:46,385 [DEBUG] argument after substitutions: 1
2019-05-24 18:19:46,386 [DEBUG] argument after substitutions: file:/
2019-05-24 18:19:46,386 [DEBUG] argument after substitutions: create,delete,link,modify
2019-05-24 18:19:46,386 [DEBUG] argument after substitutions: .*
2019-05-24 18:19:46,387 [DEBUG] sr_cpost 2.19.05b1ubuntu2 settings: action=foreground config_name=cpost_test log_level=5 follow_symlinks=no realpath=no
2019-05-24 18:19:46,387 [DEBUG] 	sleep=1 expire=180 heartbeat=300 sanity_log_dead=450 cache=0
2019-05-24 18:19:46,387 [DEBUG] 	cache_file=off accept_unmatch=on
2019-05-24 18:19:46,387 [DEBUG] 	events=000f directory=(null) queuename=(null) force_polling=off sum=d statehost=0
2019-05-24 18:19:46,387 [DEBUG] 	message_ttl=0 post_exchange=xtest post_exchange_split=0 post_exchange_suffix=(null)
2019-05-24 18:19:46,387 [DEBUG] 	source=(null) to=(null) post_base_url=file:/ topic_prefix=v02.post pid=14377
2019-05-24 18:19:46,387 [DEBUG] instance for config cpost_test (pid 14377) is not running.
2019-05-24 18:19:46,387 [DEBUG] post_broker: amqp://bunnymaster:<pw>@localhost:5672
2019-05-24 18:19:46,394 [DEBUG] 2.19.05b1ubuntu2 connected to post broker amqp://bunnymaster@localhost:5672/#xtest
2019-05-24 18:19:46,395 [DEBUG] declaring exchange amqp://bunnymaster@localhost:5672/#xtest
2019-05-24 18:19:46,396 [INFO] cpost 2.19.05b1ubuntu2 config: cpost_test, pid: 14384, starting
2019-05-24 18:19:46,396 [DEBUG] starting polling loop pass: 0
2019-05-24 18:19:46,396 [DEBUG] isMatchingPattern: /home/aspymis/cpost_test/dd matched mask: accept .*
2019-05-24 18:19:46,396 [DEBUG] info: opening directory: /home/aspymis/cpost_test/dd, first_call=on, recursive=on, follow_symlinks=off
2019-05-24 18:19:46,396 [DEBUG] info: closing directory: /home/aspymis/cpost_test/dd
2019-05-24 18:19:51,771 [DEBUG] bytes read: 32, sz ev: 32, event: 40000100 create,directory: len=16, fn=/home/aspymis/cpost_test/dd/a
2019-05-24 18:19:51,771 [DEBUG] looking in entries_done, for /home/aspymis/cpost_test/dd/a, result=(nil)
2019-05-24 18:19:51,772 [DEBUG] e->mask=40000100 from:  0040  to: 0080 
2019-05-24 18:19:51,772 [DEBUG] do one file: /home/aspymis/cpost_test/dd/a
2019-05-24 18:19:51,772 [DEBUG] isMatchingPattern: /home/aspymis/cpost_test/dd/a matched mask: accept .*
2019-05-24 18:19:51,772 [DEBUG] info: opening directory: /home/aspymis/cpost_test/dd/a, first_call=off, recursive=on, follow_symlinks=off
2019-05-24 18:19:51,772 [DEBUG] info: closing directory: /home/aspymis/cpost_test/dd/a
2019-05-24 18:20:01,979 [DEBUG] bytes read: 64, sz ev: 32, event: 0100 create: len=16, fn=/home/aspymis/cpost_test/dd/a/0
2019-05-24 18:20:01,980 [DEBUG] looking in entries_done, for /home/aspymis/cpost_test/dd/a/0, result=(nil)
2019-05-24 18:20:01,980 [DEBUG] e->mask=0100 from:  0040  to: 0080 
2019-05-24 18:20:01,980 [DEBUG] do one file: /home/aspymis/cpost_test/dd/a/0
2019-05-24 18:20:01,980 [DEBUG] isMatchingPattern: /home/aspymis/cpost_test/dd/a/0 matched mask: accept .*
2019-05-24 18:20:01,980 [DEBUG] sr_cpost file2message start with: /home/aspymis/cpost_test/dd/a/0 sb=0x7fff6079c530 islnk=0, isdir=0, isreg=1
2019-05-24 18:20:01,983 [INFO] published: 20190524182001.981157523 file:/ /home/aspymis/cpost_test/dd/a/0 topic=v02.post.home.aspymis.cpost_test.dd.a sum=d,d41d8cd98f00b204e9800998ecf8427e source=bunnymaster to_clusters=localhost from_cluster=localhost mtime=20190524182001.972 atime=20190524182001.972 mode=0664 parts=1,0,1,0,0
2019-05-24 18:20:01,985 [DEBUG] bytes read: 64, sz ev: 32, event: 0008 dunno: 0008!: len=16, fn=/home/aspymis/cpost_test/dd/a/0
2019-05-24 18:20:01,986 [DEBUG] looking in entries_done, for /home/aspymis/cpost_test/dd/a/0, result=0x55584e417b20
2019-05-24 18:20:01,986 [DEBUG] entries_done hit! ignoring:/home/aspymis/cpost_test/dd/a/0
2019-05-24 18:20:17,514 [DEBUG] bytes read: 80, sz ev: 32, event: 0200 delete: len=16, fn=/home/aspymis/cpost_test/dd/a/0
2019-05-24 18:20:17,515 [DEBUG] looking in entries_done, for /home/aspymis/cpost_test/dd/a/0, result=(nil)
2019-05-24 18:20:17,515 [DEBUG] e->mask=0200 from:  0040  to: 0080 
2019-05-24 18:20:17,515 [DEBUG] do one file: /home/aspymis/cpost_test/dd/a/0
2019-05-24 18:20:17,516 [DEBUG] sr_cpost file2message start with: /home/aspymis/cpost_test/dd/a/0 sb=(nil) islnk=0, isdir=0, isreg=0
2019-05-24 18:20:17,516 [INFO] published: 20190524182017.516489888 file:/ /home/aspymis/cpost_test/dd/a/0 topic=v02.post.home.aspymis.cpost_test.dd.a sum=R,31bca02094eb78126a517b206a88c73cfa9ec6f704c7030d18212cace820f025f00bf0ea68dbf3f3a5436ca63b53bf7bf80ad8d5de7d8359d0b7fed9dbc3ab99 source=bunnymaster to_clusters=localhost from_cluster=localhost
2019-05-24 18:20:17,517 [DEBUG] bytes read: 80, sz ev: 16, event: 8000 dunno: 8000!: len=0, fn=/home/aspymis/cpost_test/dd/a/
2019-05-24 18:20:17,517 [DEBUG] looking in entries_done, for /home/aspymis/cpost_test/dd/a/, result=(nil)
2019-05-24 18:20:17,517 [DEBUG] e->mask=8000 from:  0040  to: 0080 
2019-05-24 18:20:17,517 [DEBUG] do one file: /home/aspymis/cpost_test/dd/a/
2019-05-24 18:20:17,517 [DEBUG] isMatchingPattern: /home/aspymis/cpost_test/dd/a/ matched mask: accept .*
2019-05-24 18:20:17,517 [ERROR] file path "/home/aspymis/cpost_test/dd/a/" not utf8 encoded, ignoring sr_post call
2019-05-24 18:20:17,517 [DEBUG] bytes read: 80, sz ev: 32, event: 40000200 delete,directory: len=16, fn=/home/aspymis/cpost_test/dd/a
2019-05-24 18:20:17,517 [DEBUG] looking in entries_done, for /home/aspymis/cpost_test/dd/a, result=(nil)
2019-05-24 18:20:17,517 [DEBUG] e->mask=40000200 from:  0040  to: 0080 
2019-05-24 18:20:17,517 [DEBUG] do one file: /home/aspymis/cpost_test/dd/a
2019-05-24 18:20:17,517 [DEBUG] isMatchingPattern: /home/aspymis/cpost_test/dd/a matched mask: accept .*
2019-05-24 18:20:17,517 [DEBUG] sr_cpost file2message start with: /home/aspymis/cpost_test/dd/a sb=(nil) islnk=0, isdir=0, isreg=0
2019-05-24 18:20:17,518 [INFO] published: 20190524182017.517933468 file:/ /home/aspymis/cpost_test/dd/a topic=v02.post.home.aspymis.cpost_test.dd sum=R,1f40fc92da241694750979ee6cf582f2d5d7d28e18335de05abc54d0560e0f5302860c652bf08d560252aa5e74210546f369fbbbce8c12cfc7957b2652fe9a75 source=bunnymaster to_clusters=localhost from_cluster=localhost
2019-05-24 18:24:47,029 [DEBUG] heartbeat processing start
2019-05-24 18:24:47,036 [DEBUG] heartbeat processing completed
```

**Comment by petersilva** [comment]
We should probably deploy the new version 2.19.06b1.   Don't expect it to help, but better to track master with the problem.

**Comment by petersilva** [comment]
@msaraga This is your current highest priority, right?

**Comment by msaraga** [comment]
Definitely

**Comment by msaraga** [comment]
With `libsrshim`, we do not have the same behaviour as with `sr_cpost` when directories with content are deleted using `rm -rf`: no non-utf8 paths or headers generated

Logs:
```
aspymis@amqpbot:~/libsrshim_test/dd$ tree .
.
├── 0
├── 1
│   └── 4
└── 2
    └── 3

2 directories, 3 files
aspymis@amqpbot:~/libsrshim_test/dd$ rm -rf 1
2019-06-06 13:39:00,954 [INFO] published: 20190606133900.953635041 file:/ /home/aspymis/libsrshim_test/dd/1/4 topic=v02.post.home.aspymis.libsrshim_test.dd.1 sum=R,a321d8b405e3ef2604959847b36d171eebebc4a8941dc70a4784935a4fca5d5813de84dfa049f06549aa61b20848c1633ce81b675286ea8fb53db240d831c568 source=bunnymaster to_clusters=localhost from_cluster=localhost
2019-06-06 13:39:00,955 [INFO] published: 20190606133900.954495078 file:/ /home/aspymis/libsrshim_test/dd/1 topic=v02.post.home.aspymis.libsrshim_test.dd sum=R,4dff4ea340f0a823f15d3f4f01ab62eae0e5da579ccb851f8db9dfe84c58b2b37b89903a740e1ee172da793a6e79d560e5f7f9bd058a12a280433ed6fa46510a source=bunnymaster to_clusters=localhost from_cluster=localhost
```

Config:
```
post_broker amqp://bunnymaster@localhost/
post_exchange xtest

sleep 1

post_base_url file:/

events create,delete,link,modify

accept .*/home/aspymis/libsrshim_test/dd/.*
reject .*
```

Now, need to check what happens when we move directories

**Comment by msaraga** [comment]
most likely fixed in #39
